File: native/javaDesktopServer/src/net/mcbbs/client/main/server/cmd/PluginDashboard.java
package net.mcbbs.client.main.server.cmd;

import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

public class PluginDashboard {
    public static Set<Command> tasks = new HashSet<>();

    public static void execute(String pkg, String command, String[] args) throws IOException {
        CommandBuilder argStr = new CommandBuilder();
        for (String n : args) argStr.append(n);
        execute(pkg, command, argStr.toString());
    }

    public static void execute(String pkg, String command, String args) throws IOException {
        CommandBuilder str = new CommandBuilder();
        str.append("execute").append(pkg).append(command).append(args);
        SocketManager.sendMessage(str.toString());
        Command p = new Command(CommandType.EXECUTE, pkg, command, new String[]{args});
        tasks.add(p);
    }

    public static void data(String pkg, String command, String[] args) throws IOException {
        CommandBuilder argStr = new CommandBuilder();
        for (String n : args) argStr.append(n);
        execute(pkg, command, argStr.toString());
    }

    public static void data(String pkg, String command, String args) throws IOException {
        CommandBuilder str = new CommandBuilder();
        str.append("data").append(pkg).append(command).append(args);
        SocketManager.sendMessage(str.toString());
        Command p = new Command(CommandType.DATA, pkg, command, new String[]{args});
        tasks.add(p);
    }
}

File: native/nodeServer/src/PluginDashboard.js
class CommandBuilder{
    commands = [];
    append = str => this.commands.push(str);
    toString = () => this.commands.reduce((prev, next) => prev + " " + next).trim();
}

const dataStates = ["success", "fail"];

const PluginDashboard = {
    execute: (src, pkg, cmd, args) => {
        if(!src in sockets) throw new Error("没有名为 " + src + " 的连接！");
        let str = new CommandBuilder();
        str.append("execute");
        str.append(pkg);
        str.append(cmd);
        str.append(args);
        sendMessage(src, str.toString());
    },
    data: (src, stat, pkg, cmd, args) => {
        if(!src in sockets) throw new Error("没有名为 " + src + " 的连接！");
        if(!stat in dataStates) throw new Error("未知的状态 " + stat + "！");
        let str = new CommandBuilder();
        str.append("data");
        str.append(stat);
        str.append(pkg);
        str.append(cmd);
        str.append(args);
        sendMessage(src, str.toString());
    },
}